lib/vanity/adapters/mock_adapter.rb

```diff
@@ -134,17 +132,27 @@ def ab_not_showing(experiment, identity)
       end
 
       def ab_add_participant(experiment, alternative, identity)
-        @experiments[experiment] ||= {}
-        @experiments[experiment][:alternatives] ||= {}
-        alt = @experiments[experiment][:alternatives][alternative] ||= {}
+        alt = alternative(experiment, alternative)
         alt[:participants] ||= Set.new
         alt[:participants] << identity
       end
 
+      def ab_seen(experiment, identity, alternative)
+        if ab_assigned(experiment, identity) == alternative
+          return alternative
```

Comment:
D'oh - will correct, thanks.

Actually, I've been working on sharing the tests with the other adapters and I think I might have the method signature of this one wrong. Looking at the redis adapter it seems that for most methods having `alternative` as a parameter, they expect the alternative id. But for `ab_seen` it seems like it expects an instance of `Alternative`.

Compare `#ab_add_participant` to `#ab_seen` in the `RedisAdapter` [here](https://github.com/Futurelearn/vanity/blob/f221e16ebad0d32f74fa5c2a9ca764bcb9b1de66/lib/vanity/adapters/redis_adapter.rb#L172-L186):

```ruby
def ab_add_participant(experiment, alternative, identity)
  call_redis_with_failover(experiment, alternative, identity) do
    # interpolates `alternative` directly
    @experiments.sadd "#{experiment}:alts:#{alternative}:participants", identity
  end
end

def ab_seen(experiment, identity, alternative)
  call_redis_with_failover(experiment, identity, alternative) do
    # calls `#id` before interpolating
    if @experiments.sismember "#{experiment}:alts:#{alternative.id}:participants", identity
      alternative
    else
      nil
    end
  end
end
```

Is that right? If so I'll update the test and implementation...